File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n
  | n <= 0 = Nothing
  | otherwise = Just $ decide n ((sum . decompose) n)
  where
    decide x d 
      | x < d = Abundant
      | x == d = Perfect
      | x > d = Deficient
  
decompose :: Int -> [Int]
decompose n = [i | i <- [1..(n-1)], rem n i == 0]

File: python/perfect-numbers/perfect_numbers.py
def classify(number):

    if number in [0,-1]:
        raise ValueError("Must provide a non-zero natural number")
    
    s = 0
    for n in range(1, number):
        if number % n == 0:
            s += n
            
    if s == number:
        return "perfect"
    if s > number:
        return "abundant"
    if s < number:
        return "deficient"
